File: Task/Arithmetic-geometric-mean-Calculate-Pi/Scala/arithmetic-geometric-mean-calculate-pi.scala
import java.math.MathContext

import scala.annotation.tailrec
import scala.compat.Platform.currentTime
import scala.math.BigDecimal

object Calculate_Pi extends App {
  val precision = new MathContext(32768 /*65536*/)
  val (bigZero, bigOne, bigTwo, bigFour) =
    (BigDecimal(0, precision), BigDecimal(1, precision), BigDecimal(2, precision), BigDecimal(4, precision))

  def bigSqrt(bd: BigDecimal) = {
    @tailrec
    def iter(x0: BigDecimal, x1: BigDecimal): BigDecimal =
      if (x0 == x1) x1 else iter(x1, (bd / x1 + x1) / bigTwo)

    iter(bigZero, BigDecimal(Math.sqrt(bd.toDouble), precision))
  }

  @tailrec
  private def loop(a: BigDecimal, g: BigDecimal, sum: BigDecimal, pow: BigDecimal): BigDecimal = {
    if (a == g) (bigFour * (a * a)) / (bigOne - sum)
    else {
      val (_a, _g, _pow) = ((a + g) / bigTwo, bigSqrt(a * g), pow * bigTwo)
      loop(_a, _g, sum + ((_a * _a - (_g * _g)) * _pow), _pow)
    }
  }

  println(precision)
  val pi = loop(bigOne, bigOne / bigSqrt(bigTwo), bigZero, bigTwo)
  println(s"This are ${pi.toString.length - 1} digits of π:")
  val lines = pi.toString().sliding(103, 103).mkString("\n")
  println(lines)

  println(s"Successfully completed without errors. [total ${currentTime - executionStart} ms]")
}

File: Task/Arithmetic-geometric-mean-Calculate-Pi/JavaScript/arithmetic-geometric-mean-calculate-pi.js
const Decimal = require('decimal.js');

// Set global precision (e.g., 512 digits)
Decimal.set({ precision: 512 });

const bigTwo = new Decimal(2);
const bigFour = new Decimal(4);

function bigSqrt(bd) {
    let x0 = new Decimal(0);
    let x1 = new Decimal(Math.sqrt(bd.toNumber()));
    while (!x0.equals(x1)) {
        x0 = x1;
        x1 = bd.div(x0).plus(x0).div(bigTwo);
    }
    return x1;
}

function calculatePi() {
    let a = new Decimal(1);
    let g = a.div(bigSqrt(bigTwo));
    let t;
    let sum = new Decimal(0);
    let pow = bigTwo;

    while (!a.equals(g)) {
        t = a.plus(g).div(bigTwo);
        g = bigSqrt(a.times(g));
        a = t;
        pow = pow.times(bigTwo);
        sum = sum.plus(a.times(a).minus(g.times(g)).times(pow));
    }

    const pi = bigFour.times(a.times(a)).div(new Decimal(1).minus(sum));
    console.log(pi.toString());
}

calculatePi();
